### PR TITLE
feat: Calendar widget for publicly shared calendars

### PR DESCRIPTION
**Pull request by hamza221**
For #5746

## How to test
1. Create a new calendar
2. Add events to the calendar
3. Create a share link for the calendar (URL includes the /p/ part)
4. Open Talk or Text
5. Paste the link

image

### CODE REVIEW

File: lib/Reference/ReferenceProvider.php [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+	 * @inheritDoc
+	 */
+	public function getIconUrl(): string {
+		return $this->urlGenerator->getAbsoluteURL(
+			$this->urlGenerator->imagePath(Application::APP_ID, 'calendar-dark.svg')
+		);
+	}
+
+	public function matchReference(string $referenceText): bool {
+		$start = $this->urlGenerator->getAbsoluteURL('/index.php/apps/' . Application::APP_ID);
```
Comment (ChristophWurst) [inline lib/Reference/ReferenceProvider.php:74]: - [ ] Does not work for instances with pretty URLs

**Review by ChristophWurst — COMMENTED** [review]
Tested and basics work :+1:

* No event details popover menu on click
* Hide export/download action
* Hide subscribe action
* Styling

**Review by st3iny — COMMENTED** [review]
Year and list views look weird on (Nextcloud's) dark theme.

| Year | List |
| --- | --- |
| Screenshot_20240318_133503 | Screenshot_20240318_133508 |

**Review by st3iny — COMMENTED** [review]
- [x] Clicking on events does not work. Is that supposed to work and show a popover?

**EDIT:** Looking at the code it is not supposed to work, so feel free to tick the box.

File: src/components/AppNavigation/AppNavigationHeader/AppNavigationHeaderDatePicker.vue [hunk 2]
```diff
@@ -92,7 +102,7 @@ export default {
 			locale: (state) => state.settings.momentLocale,
 		}),
 		selectedDate() {
-			return getDateFromFirstdayParam(this.$route.params?.firstDay ?? 'now')
+			return this.isWidget ? getDateFromFirstdayParam(this.$store.getters.widgetDate) : getDateFromFirstdayParam(this.$route.params?.firstDay ?? 'now')
```
Comment (st3iny) [inline src/components/AppNavigation/AppNavigationHeader/AppNavigationHeaderDatePicker.vue:105]: Nit: Refactor ternary operator

File: src/components/AppNavigation/AppNavigationHeader/AppNavigationHeaderDatePicker.vue [hunk 3]
```diff
@@ -139,7 +149,7 @@ export default {
 			}
 		},
 		view() {
-			return this.$route.params.view
+			return this.isWidget ? this.$store.getters.widgetView : this.$route?.params?.view
```
Comment (st3iny) [inline src/components/AppNavigation/AppNavigationHeader/AppNavigationHeaderDatePicker.vue:152]: Nit: Refactor ternary operator

File: src/components/AppNavigation/EmbedHeader/EmbedHeaderViewButtons.vue [hunk 4]
```diff
@@ -76,22 +82,27 @@ export default {
 			return this.selectedView === 'listMonth'
 		},
 		selectedView() {
-			return this.$route.params.view
+			return this.isWidget ? this.$store.getters.widgetView : this.$route?.params?.view
```
Comment (st3iny) [inline src/components/AppNavigation/EmbedHeader/EmbedHeaderViewButtons.vue:85]: Nit: Refactor ternary operator

File: src/views/Calendar.vue [hunk 5]
```diff
@@ -152,29 +171,30 @@ export default {
...
-			return !this.isPublicShare && !this.isEmbedded
+			return !this.isPublicShare && !this.isEmbedded && !this.isWidget
 		},
 		isAuthenticatedUser() {
-			return !this.isPublicShare && !this.isEmbedded
+			return !this.isPublicShare && !this.isEmbedded && !this.isWidget
 		},
 		isPublicShare() {
-			return this.$route.name.startsWith('Public')
+			return this.isWidget ? false : this.$route?.name.startsWith('Public')
```
Comment (st3iny) [inline src/views/Calendar.vue:191]: Nit: Refactor ternary operator

**Review by st3iny — COMMENTED** [review]
Tested and works. Code looks good to me.

Left some feedback above and below.

Nitpick: I'm not a fan of ternaries if they are not necessary (e.g. when used in returns) and multiline if statements are easier to read and follow. Addressing my nitpicks is not required.

**Comment by ChristophWurst** [comment]
> **EDIT:** Looking at the code it is not supposed to work, so feel free to tick the box.

It should work IMO. The idea was to have something similar to calendars embedded into iframes. You can see the details of events there.

File: lib/Listener/CalendarReferenceListener.php [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+		$tasksEnabled = $this->appManager->isEnabledForUser('tasks');
+
+		$circleVersion = $this->appManager->getAppVersion('circles');
+		$isCirclesEnabled = $this->appManager->isEnabledForUser('circles') === true;
+		// if circles is not installed, we use 0.0.0
+		$isCircleVersionCompatible = $this->compareVersion->isCompatible($circleVersion ? $circleVersion : '0.0.0', '22');
+
+		$this->initialStateService->provideInitialState('app_version', $appVersion);
+		$this->initialStateService->provideInitialState('event_limit', $defaultEventLimit);
+		$this->initialStateService->provideInitialState('first_run', true);
```
Comment (tcitworld) [inline lib/Listener/CalendarReferenceListener.php:107]:
```suggestion
		$this->initialStateService->provideInitialState('first_run', false);
```
Fun fact, it seems this config isn't even used anymore.

File: lib/Listener/CalendarReferenceListener.php [hunk 7]
```diff
@@ -0,0 +1,130 @@
...
+		$this->initialStateService->provideInitialState('attachments_folder', '/Calendar');
+		$this->initialStateService->provideInitialState('slot_duration', $defaultSlotDuration);
+		$this->initialStateService->provideInitialState('default_reminder', $defaultDefaultReminder);
+		$this->initialStateService->provideInitialState('tasks_enabled', $tasksEnabled);
+		$this->initialStateService->provideInitialState('hide_event_export', true);
+		$this->initialStateService->provideInitialState('force_event_alarm_type', $forceEventAlarmType);
+		$this->initialStateService->provideInitialState('disable_appointments', true);
+		$this->initialStateService->provideInitialState('can_subscribe_link', false);
+		$this->initialStateService->provideInitialState('show_resources', $showResources);
+		$this->initialStateService->provideInitialState('isCirclesEnabled', $isCirclesEnabled && $isCircleVersionCompatible);
```
Comment (tcitworld) [inline lib/Listener/CalendarReferenceListener.php:125]: No need for that

File: lib/Listener/CalendarReferenceListener.php [hunk 8]
```diff
@@ -0,0 +1,130 @@
...
+		$isCircleVersionCompatible = $this->compareVersion->isCompatible($circleVersion ? $circleVersion : '0.0.0', '22');
+
+		$this->initialStateService->provideInitialState('app_version', $appVersion);
+		$this->initialStateService->provideInitialState('event_limit', $defaultEventLimit);
+		$this->initialStateService->provideInitialState('first_run', true);
+		$this->initialStateService->provideInitialState('initial_view', $defaultInitialView);
+		$this->initialStateService->provideInitialState('show_weekends', $defaultShowWeekends);
+		$this->initialStateService->provideInitialState('show_week_numbers', $defaultWeekNumbers === 'yes');
+		$this->initialStateService->provideInitialState('skip_popover', true);
+		$this->initialStateService->provideInitialState('talk_enabled', true);
```
Comment (tcitworld) [inline lib/Listener/CalendarReferenceListener.php:112]:
```suggestion
		$this->initialStateService->provideInitialState('talk_enabled', false);
```

**Review by tcitworld — COMMENTED** [review]
Generally speaking just use what's set in `PublicViewController`.

https://github.com/nextcloud/calendar/blob/ffa02bded4c05099ea6ee85219252a678af76677/lib/Controller/PublicViewController.php#L112-L148

**Comment by hamza221** [comment]
> Tested and basics work 👍
> 
> * No event details popover menu on click
> * Hide export/download action
> * Hide subscribe action
> * Styling

I think all done ✅

**Review by GretaD — APPROVED** [review]
it works

**Comment by ChristophWurst** [comment]
Looks good

image

Minor remarks
1. Export is still shown
2. Subscribe is still shown
3. Actions in the toolbar are not aligned
4. Date input has unbalanced top/left padding
5. Scrollbar on the right necessary?
6. Space between right border of the calendar and the widget is weird

**Comment by hamza221** [comment]
> Looks good
> 
> image
> 
> Minor remarks
> 
> 1. Export is still shown
> 2. Subscribe is still shown
> 3. Actions in the toolbar are not aligned
> 4. Date input has unbalanced top/left padding
> 5. Scrollbar on the right necessary?
> 6. Space between right border of the calendar and the widget is weird

are you on the latest commit ,  1, 2 should be already solved
image

**Comment by ChristophWurst** [comment]
I am on 35721533be079bff09a02f8d5535be4958e18cc9

**Comment by ChristophWurst** [comment]
Indeed resolved. I forgot to test with cleared caches.

**Review by ChristophWurst — APPROVED** [review]
Works

Let's add unit tests for the two new PHP classes in a separate PR

File: src/fullcalendar/interaction/eventClick.js [hunk 9]
```diff
@@ -36,17 +36,23 @@ import { emit } from '@nextcloud/event-bus'
...
+ * @param {boolean} isWidget Whether the calendar is embedded in a widget
+ * @param {object} widgetRef
  * @return {Function}
  */
-export default function(store, router, route, window) {
+export default function(store, router, route, window, isWidget = false, widgetRef = undefined) {
+
 	return function({ event }) {
+		if (isWidget) {
+			store.commit('setWidgetRef', { widgetRef: widgetRef.$el })
```
[the remark this answers is not in the document]
Comment (hamza221) [reply src/fullcalendar/interaction/eventClick.js:47]: `eventClick: eventClick(this.$store, this.$router, this.$route, window, this.isWidget, this.$refs.fullCalendar),`

The ref doesn't exist yet when initialising options

[the remark this answers is not in the document]
Comment (hamza221) [reply src/fullcalendar/interaction/eventClick.js:47]: Fix in #5865

**Pull request merged**